Doc from openstack admin guide (#1083)
stackhpc/stackhpc-kayobe-config: doc/source/operations/control-plane-operation.rst

with the following command:

.. code-block:: console

   kayobe# kayobe overcloud database recover

Ansible Control Host
--------------------

The Ansible control host is not enrolled in Bifrost and will have to be powered
on manually.

Seed VM
-------

The seed VM (and any other service VM) should start automatically when the seed
hypervisor is powered on. If it does not, it can be started with:

.. code-block:: console

   kayobe# virsh start <Seed hostname>

Full power on
-------------

Follow the order in :ref:`full-shutdown`, but in reverse order.

Shutting Down / Restarting Monitoring Services
----------------------------------------------

Shutting down
+++++++++++++

Log into the monitoring host(s):

.. code-block:: console

   kayobe# ssh stack@monitoring0

Stop all Docker containers:

.. code-block:: console

   monitoring0# for i in `docker ps -q`; do docker stop $i; done

Shut down the node:

.. code-block:: console

   monitoring0# sudo shutdown -h

Starting up
+++++++++++

The monitoring services containers will automatically start when the monitoring
node is powered back on.

Software Updates
================

> Update Packages on Control Plane

Reviewer: This should reference our release train page, because it misses all the steps for syncing pulp content

--------------------------------

OS packages can be updated with:

.. code-block:: console

   kayobe# kayobe overcloud host package update --limit <Hypervisor node> --packages '*'
   kayobe# kayobe overcloud seed package update --packages '*'

See https://docs.openstack.org/kayobe/latest/administration/overcloud.html#updating-packages

Minor Upgrades to OpenStack Services
------------------------------------

* Pull latest changes from upstream stable branch to your own ``kolla`` fork (if applicable)
* Update ``kolla_openstack_release`` in ``etc/kayobe/kolla.yml`` (unless using default)
* Update tags for the images in ``etc/kayobe/kolla/globals.yml`` to use the new value of ``kolla_openstack_release``
* Rebuild container images
* Pull container images to overcloud hosts

> * Run kayobe overcloud service upgrade

Reviewer: These instructions are much too simplified. I think it would be best to remove this section and just link to the upgrading.rst page.

For more information, see: https://docs.openstack.org/kayobe/latest/upgrading.html

Troubleshooting
===============

Deploying to a Specific Hypervisor
----------------------------------

To test creating an instance on a specific hypervisor, *as an admin-level user*
you can specify the hypervisor name as part of an extended availability zone
description.

To see the list of hypervisor names:

.. code-block:: console

   # From host that can reach Openstack
   openstack hypervisor list

To boot an instance on a specific hypervisor

.. code-block:: console

>    openstack server create --flavor <Flavour name>--network <Network name> --key-name <key> --image <Image name> --availability-zone nova::<Hypervisor name> <VM name>

Reviewer: Should use `--os-compute-api-version 2.74 --host <hostname>` rather than availability zones

overview of the key things you need to backup:

Controllers
-----------

* `Back up SQL databases <https://docs.openstack.org/kayobe/latest/administration/overcloud.html#performing-database-backups>`__
* `Back up configuration in /etc/kolla <https://docs.openstack.org/kayobe/latest/administration/overcloud.html#saving-overcloud-service-configuration>`__

Compute
-------

The compute nodes can largely be thought of as ephemeral, but you do need to
make sure you have migrated any instances and disabled the hypervisor before
rebooting, decommissioning or making any disruptive configuration change.

Monitoring
----------

* `Back up InfluxDB <https://docs.influxdata.com/influxdb/v1.8/administration/backup_and_restore/>`__
* `Back up OpenSearch <https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/>`__
* `Back up Prometheus <https://prometheus.io/docs/prometheus/latest/querying/api/#snapshot>`__

Seed
----

* `Back up bifrost <https://docs.openstack.org/kayobe/latest/administration/seed.html#database-backup-restore>`__

Ansible control host
--------------------

* Back up service VMs such as the seed VM

Control Plane Monitoring

> ========================

Reviewer: Could you add a link in this section to `monitoring.rst` which has more information on how to configure the services?

Author: ACK

such as the seed virtual machine which will need to be gracefully powered down.

Compute
-------

If you are shutting down a single hypervisor, to avoid down time to tenants it
is advisable to migrate all of the instances to another machine. See
:ref:`evacuating-all-instances`.

Ceph
----

The following guide provides a good overview:
https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/8/html/director_installation_and_usage/sect-rebooting-ceph

Shutting down the seed VM
-------------------------

.. code-block:: console

   kayobe# virsh shutdown <Seed hostname>

.. _full-shutdown:

Full shutdown
-------------

In case a full shutdown of the system is required, we advise to use the
following order:

* Perform a graceful shutdown of all virtual machine instances
* Shut down compute nodes
* Shut down monitoring node (if separate from controllers)
* Shut down network nodes (if separate from controllers)
* Shut down controllers
* Shut down Ceph nodes (if applicable)
* Shut down seed VM
* Shut down Ansible control host

Rebooting a node
----------------

Example: Reboot all compute hosts apart from compute0:

.. code-block:: console

>    kayobe# kayobe overcloud host command run --limit 'compute:!compute0' -b --command "shutdown -r"

Reviewer: We should direct users to run the reboot playbook instead, since it covers a few more edge cases than this

custom rules.

Silencing Prometheus Alerts
---------------------------

Sometimes alerts must be silenced because the root cause cannot be resolved
right away, such as when hardware is faulty. For example, an unreachable
hypervisor will produce several alerts:

* ``InstanceDown`` from Node Exporter
* ``OpenStackServiceDown`` from the OpenStack exporter, which reports status of
  the ``nova-compute`` agent on the host
* ``PrometheusTargetMissing`` from several Prometheus exporters

Rather than silencing each alert one by one for a specific host, a silence can
apply to multiple alerts using a reduced list of labels. Log into Alertmanager,
click on the ``Silence`` button next to an alert and adjust the matcher list
to keep only ``instance=<hostname>`` label.
Then, create another silence to match ``hostname=<hostname>`` (this is
required because, for the OpenStack exporter, the instance is the host running
the monitoring service rather than the host being monitored).

.. note::

   After creating the silence, you may get redirected to a 404 page. This is a
   `known issue <https://github.com/prometheus/alertmanager/issues/1377>`__

>    when running several Alertmanager instances behind HAProxy.

Reviewer: I've never seen this before. Might be resolved?

Author: I agree

and the other:

.. code-block:: console

   VRRP_Instance(kolla_internal_vip_51) Entering MASTER STATE

Ansible Control Host
--------------------

The Ansible control host is not enrolled in bifrost. This node may run services
such as the seed virtual machine which will need to be gracefully powered down.

Compute
-------

If you are shutting down a single hypervisor, to avoid down time to tenants it
is advisable to migrate all of the instances to another machine. See
:ref:`evacuating-all-instances`.

Ceph
----

The following guide provides a good overview:
https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/8/html/director_installation_and_usage/sect-rebooting-ceph

Shutting down the seed VM
-------------------------

.. code-block:: console

   kayobe# virsh shutdown <Seed hostname>

.. _full-shutdown:

Full shutdown
-------------

In case a full shutdown of the system is required, we advise to use the
following order:

* Perform a graceful shutdown of all virtual machine instances
* Shut down compute nodes
* Shut down monitoring node (if separate from controllers)
* Shut down network nodes (if separate from controllers)
* Shut down controllers

> * Shut down Ceph nodes (if applicable)

Reviewer: Controllers before Ceph nodes seems weird, please check this with someone who knows Ceph

* Shut down seed VM
* Shut down Ansible control host

Rebooting a node
----------------

Use ``reboot.yml`` playbook to reboot nodes
Example: Reboot all compute hosts apart from compute0:

.. code-block:: console

>    kayobe# kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/reboot.yml --limit 'compute:!compute0'

Reviewer: The limit on this command is unnecessarily weird and complicated. I'd change it to something a bit easier to understand
```suggestion
   kayobe# kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/reboot.yml --limit compute4,compute5
```

Author: That is valid syntax and I think it's a necessary example.
Users might want to try rebooting all hypervisors except for a few.
If they have large number of hypervisors it would be painful to write every hypervisors to reboot with comma separated list.

environment, you can promote them by:

.. code-block:: console

   kayobe# kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/pulp-repo-promote-production.yml

To sync container images:

.. code-block:: console

   kayobe# kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/pulp-container-sync.yml
   kayobe# kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/pulp-container-publish.yml

For more information about StackHPC Release Train, see :ref:`stackhpc-release-train` documentation.

Once sync with StackHPC Release Train is done, new contents will be accessible from local
Pulp server.

Update Host Packages on Control Plane
-------------------------------------

Host packages can be updated with:

.. code-block:: console

   kayobe# kayobe overcloud host package update --limit <node> --packages '*'
   kayobe# kayobe seed host package update --packages '*'

See https://docs.openstack.org/kayobe/latest/administration/overcloud.html#updating-packages

Troubleshooting
===============

Deploying to a Specific Hypervisor
----------------------------------

To test creating an instance on a specific hypervisor, *as an admin-level user*

> you can specify the hypervisor name as part of an extended availability zone
> description.

Reviewer: The availability zone trick is the old deprecated way of doing it
```suggestion
you can specify the hypervisor name.
```

=======================
Operating Control Plane
=======================

Backup of the OpenStack Control Plane
=====================================

As the backup procedure is constantly changing, it is normally best to check
the upstream documentation for an up to date procedure. Here is a high level
overview of the key things you need to backup:

Controllers
-----------

* `Back up SQL databases <https://docs.openstack.org/kayobe/latest/administration/overcloud.html#performing-database-backups>`__
* `Back up configuration in /etc/kolla <https://docs.openstack.org/kayobe/latest/administration/overcloud.html#saving-overcloud-service-configuration>`__

Compute
-------

The compute nodes can largely be thought of as ephemeral, but you do need to
make sure you have migrated any instances and disabled the hypervisor before
decommissioning or making any disruptive configuration change.

Monitoring
----------

* `Back up InfluxDB <https://docs.influxdata.com/influxdb/v1.8/administration/backup_and_restore/>`__

> * `Back up ElasticSearch <https://www.elastic.co/guide/en/elasticsearch/reference/current/backup-cluster-data.html>`__

Reviewer: ```suggestion
* `Back up OpenSearch <https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/>`__
```

* `Back up Prometheus <https://prometheus.io/docs/prometheus/latest/querying/api/#snapshot>`__

Seed
----

* `Back up bifrost <https://docs.openstack.org/kayobe/latest/administration/seed.html#database-backup-restore>`__

Ansible control host
--------------------

* Back up service VMs such as the seed VM

Control Plane Monitoring
========================

> The control plane has been configured to collect logs centrally using the EFK
> stack (Elasticsearch, Fluentd and Kibana).

Reviewer: ```suggestion
The control plane has been configured to collect logs centrally using
OpenSearch, Fluentd, and OpenSearch Dashboards).
```

* Bifrost should automatically power on the node via IPMI
* Check that all docker containers are running
* Check Kibana for any messages with log level ERROR or equivalent

Controllers
-----------

If all of the servers were shut down at the same time, it is necessary to run a
script to recover the database once they have all started up. This can be done
with the following command:

.. code-block:: console

   kayobe# kayobe overcloud database recover

Ansible Control Host
--------------------

The Ansible control host is not enrolled in Bifrost and will have to be powered
on manually.

Seed VM
-------

The seed VM (and any other service VM) should start automatically when the seed
hypervisor is powered on. If it does not, it can be started with:

.. code-block:: console

   kayobe# virsh start <Seed hostname>

Full power on
-------------

Follow the order in :ref:`full-shutdown`, but in reverse order.

Shutting Down / Restarting Monitoring Services
----------------------------------------------

Shutting down
+++++++++++++

Log into the monitoring host(s):

.. code-block:: console

   kayobe# ssh stack@monitoring0

Stop all Docker containers:

.. code-block:: console

>    monitoring0# for i in `docker ps -q`; do docker stop $i; done

Reviewer: Should use `systemctl` rather than docker

   kayobe# kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/pulp-container-sync.yml
   kayobe# kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/pulp-container-publish.yml

For more information about StackHPC Release Train, see :ref:`stackhpc-release-train` documentation.

Once sync with StackHPC Release Train is done, new contents will be accessible from local
Pulp server.

Update Host Packages on Control Plane
-------------------------------------

Host packages can be updated with:

.. code-block:: console

   kayobe# kayobe overcloud host package update --limit <node> --packages '*'
   kayobe# kayobe seed host package update --packages '*'

See https://docs.openstack.org/kayobe/latest/administration/overcloud.html#updating-packages

Upgrading OpenStack Services
----------------------------

* Update tags for the images in ``etc/kayobe/kolla-image-tags.yml``
* Pull container images to overcloud hosts with ``kayobe overcloud container image pull``
* Run ``kayobe overcloud service upgrade``

You can update the subset of containers or hosts by

.. code-block:: console

   kayobe# kayobe overcloud service upgrade --kolla-tags <service> --limit <hostname> --kolla-limit <hostname>

> For more information, see: https://docs.openstack.org/kayobe/latest/upgrading.html

Reviewer: I'd probably just delete this section. Full version upgrades require a lot more than this. For container tag bumps, we just use service deploy. Either way, I don't think this section is really that useful

Troubleshooting
===============

Deploying to a Specific Hypervisor
----------------------------------

To test creating an instance on a specific hypervisor, *as an admin-level user*
you can specify the hypervisor name as part of an extended availability zone
description.

To see the list of hypervisor names:

.. code-block:: console

   # From host that can reach Openstack
   openstack hypervisor list

To boot an instance on a specific hypervisor

.. code-block:: console

>    openstack server create --flavor <flavour name>--network <network name> --key-name <key> --image <Image name> --os-compute-api-version 2.74 --host <hypervisor hostname> <vm name>

Reviewer: ```suggestion
   openstack server create --flavor <flavour name> --network <network name> --key-name <key> --image <Image name> --os-compute-api-version 2.74 --host <hypervisor hostname> <vm name>
```

> Cleanup Procedures

Reviewer: Patch merged and backported long ago. This can be removed
